db/hints/internal/hint_sender.cc

```diff
@@ -80,9 +80,20 @@ future<> hint_sender::do_send_one_mutation(frozen_mutation_and_schema m, locator
     return futurize_invoke([this, m = std::move(m), ermp = std::move(ermp), &natural_endpoints] () mutable -> future<> {
         // The fact that we send with CL::ALL in both cases below ensures that new hints are not going
         // to be generated as a result of hints sending.
-        if (boost::range::find(natural_endpoints, end_point_key()) != natural_endpoints.end()) {
+        const auto maybe_addr = [&] () -> std::optional<gms::inet_address> {
+            const auto& tm = ermp->get_token_metadata();
+            if (auto maybe_ep = tm.get_endpoint_for_host_id_if_known(end_point_key())) {
+                return *maybe_ep;
+            }
+            if (tm.get_topology().is_me(end_point_key())) {
+                return tm.get_topology().my_address();
+            }
+            return {};
+        } ();
```

Comment:
Style note: I prefer std::invoke(long lambda) to (long lambda)(). The former syntax immediately tells me I'm looking at a lambda introduced for a scope, while the latter syntax keeps me thinking we're defining a helper until we see the invocation (or don't see it since it's just three characters).